flink-connector-aws-kinesis-streams/src/main/java/org/apache/flink/connector/kinesis/source/enumerator/KinesisStreamsSourceEnumerator.java

```diff
@@ -0,0 +1,338 @@
+    /**
+     * This method is used to discover Kinesis splits the job can subscribe to. It can be run in
+     * parallel, is important to not mutate any shared state.
+     *
+     * @return list of discovered splits
+     */
+    private List<KinesisShardSplit> periodicallyDiscoverSplits() {
+        List<Shard> shards = streamProxy.listShards(streamArn, lastSeenShardId);
+        // Any shard discovered after the initial startup should be read from the start, since they
+        // come from resharding
+        return mapToSplits(shards, true);
+    }
+
+    private List<KinesisShardSplit> mapToSplits(List<Shard> shards, boolean shouldReadFromStart) {
+        InitialPosition initialPositionFromConfig =
+                shouldReadFromStart
+                        ? InitialPosition.TRIM_HORIZON
+                        : InitialPosition.valueOf(
+                                consumerConfig
+                                        .getOrDefault(
+                                                STREAM_INITIAL_POSITION,
+                                                DEFAULT_STREAM_INITIAL_POSITION)
+                                        .toString());
+        StartingPosition startingPosition;
+        switch (initialPositionFromConfig) {
+            case LATEST:
+                // If LATEST is requested, we still set the starting position to the time of
+                // startup. This way, the job starts reading from a deterministic timestamp
+                // (i.e. time of job submission), even if it enters a restart loop immediately
+                // after submission.
+                startingPosition = StartingPosition.fromTimestamp(Instant.now());
+                break;
+            case AT_TIMESTAMP:
+                startingPosition =
+                        StartingPosition.fromTimestamp(
+                                parseStreamTimestampStartingPosition(consumerConfig).toInstant());
+                break;
+            case TRIM_HORIZON:
+            default:
+                startingPosition = StartingPosition.fromStart();
+        }
+
+        List<KinesisShardSplit> splits = new ArrayList<>();
+        for (Shard shard : shards) {
+            splits.add(new KinesisShardSplit(streamArn, shard.shardId(), startingPosition));
+        }
+
+        return splits;
+    }
+
+    /**
+     * This method assigns a given set of Kinesis splits to the readers currently registered on the
+     * cluster. This assignment is done via a side-effect on the {@link SplitEnumeratorContext}
+     * object.
+     *
+     * @param discoveredSplits list of discovered splits
+     * @param t throwable thrown when discovering splits. Will be null if no throwable thrown.
+     */
+    private void assignSplits(List<KinesisShardSplit> discoveredSplits, Throwable t) {
+        if (t != null) {
+            throw new KinesisStreamsSourceException("Failed to list shards.", t);
+        }
+
+        if (context.registeredReaders().isEmpty()) {
```

Comment:
Hm, to be honest I don't expect this to be ever hit in our scenario, but I put this in because the lifecycle of `listShard()` is on JM, and lifecycle of `registerReader` is on TM, so we don't have a strong guarantee that we will always registerReader before listShard. 

Also, on partial failover, the reader that experienced failover is removed.  If we support partial failover, and all subtasks fail partially, we could end up with no registered readers.


In this case, since we can put it into unassignedSplits, we should be ok to handle this transparently.

Reply:
Following on from this, it seems if there is at least 1 reader we will blindly assign splits to it based on the shard assigner output. I wonder if we should skip assignment when the reader is not registered? Another rejected option is to assign all splits to the registered readers, but this will result in skew so not a good idea.

Reply:
Added a check to only assign splits when there are enough readers